hibernate-envers/src/main/java/org/hibernate/envers/query/internal/impl/RevisionsOfEntityAssociationQuery.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.envers.query.internal.impl;
+
+import java.util.Map;
+
+import javax.persistence.criteria.JoinType;
+
+import org.hibernate.Incubating;
+import org.hibernate.envers.boot.internal.EnversService;
+import org.hibernate.envers.configuration.internal.AuditEntitiesConfiguration;
+import org.hibernate.envers.internal.entities.EntityConfiguration;
+import org.hibernate.envers.internal.entities.mapper.id.IdMapper;
+import org.hibernate.envers.internal.reader.AuditReaderImplementor;
+import org.hibernate.envers.internal.tools.query.Parameters;
+import org.hibernate.envers.internal.tools.query.QueryBuilder;
+import org.hibernate.envers.query.AuditAssociationQuery;
+
+/**
+ * An {@link org.hibernate.envers.query.AuditAssociationQuery} implementation for {@link RevisionsOfEntityQuery}.
+ *
+ * @author Chris Cranford
+ */
+@Incubating
+public class RevisionsOfEntityAssociationQuery<Q extends AuditQueryImplementor> extends AbstractAuditAssociationQuery<Q> {
+
+    public RevisionsOfEntityAssociationQuery(
+            EnversService enversService,
+            AuditReaderImplementor auditReader,
+            Q parent,
+            QueryBuilder queryBuilder,
+            String propertyName,
+            JoinType joinType,
+            Map<String, String> aliasToEntityNameMap,
+            String ownerAlias,
+            String userSuppliedAlias) {
+        super( enversService,
+               auditReader,
+               parent,
+               queryBuilder,
+               propertyName,
+               joinType,
+               aliasToEntityNameMap,
+               ownerAlias,
+               userSuppliedAlias );
+    }
+
+    @Override
+    public AuditAssociationQuery<? extends AuditAssociationQuery<Q>> traverseRelation(
+            String associationName,
+            JoinType joinType,
+            String alias) {
+        AbstractAuditAssociationQuery<AbstractAuditAssociationQuery<Q>> result = associationQueryMap.get( associationName );
+        if ( result == null ) {
+            result = new RevisionsOfEntityAssociationQuery<>(
+                    enversService,
+                    auditReader,
+                    this,
+                    queryBuilder,
+                    associationName,
+                    joinType,
+                    aliasToEntityNameMap,
+                    this.alias,
+                    alias );
+
+            associationQueries.add( (RevisionsOfEntityAssociationQuery<Q>) result );
+            associationQueryMap.put( associationName, result );
+        }
```

Comment:
The above code pattern involves a duplicated down casting, which can be avoided as follows:
```
if ( !associationQueryMap.containsKey( associationName ) ) {
    RevisionsOfEntityAssociationQuery<Q> query = new RevisionsOfEntityAssociationQuery<>(
                    enversService,
                    auditReader,
                    this,
                    queryBuilder,
                    associationName,
                    joinType,
                    aliasToEntityNameMap,
                    this.alias,
                    alias );
    associationQueries.add( query );
    associationQueryMap.put( associationName, query );
}
```
seems a little bit elegant, though not a big deal as well.

You might well code as follows:
```
associationQueryMap.computeIfAbsent( associationName, name -> {
    final RevisionsOfEntityAssociationQuery<Q> query = new RevisionsOfEntityAssociationQuery<>(
                    enversService,
                    auditReader,
                    this,
                    queryBuilder,
                    associationName,
                    joinType,
                    aliasToEntityNameMap,
                    this.alias,
                    alias );
    associationQueries.add( query );
    return query;
} );
```
though not everybody is big fan of it.

Reply:
I believe the `computeIfAbsent` compute function throws an argument inference compilation failure here, you either have issues with the `<>` inference or the reference to `this` in the constructor argument list.